Add docs for pullsecrets in CatalogSources

adellape opened the pull request:
https://issues.redhat.com/browse/OSDOCS-1708

Preview (internal): http://file.rdu.redhat.com/~adellape/012021/olm_pullsecret/operators/admin/olm-managing-custom-catalogs.html#olm-accessing-images-private-registries_olm-managing-custom-catalogs

[skip netlify]

jeana-redhat reviewed (approved):
looks great!

anik120 commented on modules/olm-adding-pullsecret-catalogsource.adoc:
```diff
@@ -0,0 +1,46 @@
+[id="olm-adding-pullsecret-catalogsource_{context}"]
+= Adding pull secrets to catalog sources
+
+If an index image or catalog image is hosted on an authenticated container image registry, also known as a private registry, you can create a pull secret that contains the authentication credentials for the registry. By referencing the secret in a catalog source, the required images can be pulled for use in OperatorHub.
```
Probably should mention that if any operator bundle inside the catalog is hosted in an authenticated registry, i.e a secret is needed to access the bundle image, the secret should be passed on via `spec.secrets` for CatalogSource. 
If there are two operators hosted in different private registries, and requires two secrets to access two bundles, both secrets should be passed onto `sepc.Secrets`

anik120 reviewed (commented):
Looking great @adellape 
Just had a suggestion.

jianzhangbjz commented on modules/olm-adding-pullsecret-catalogsource.adoc:
```diff
@@ -0,0 +1,46 @@
+[source,terminal]
+----
+$ oc create secret generic <secret_name> \
+    --from-file=.dockerconfigjson=/root/.docker/config.json \
```
If using this `/root/.docker/config.json` file, we'd better add a login step to get the private auth info, such as:
```console
$ docker login <registry>:<port>
```
Or we can replace this  `/root/.docker/config.json` file with other auth files.

jianzhangbjz commented on modules/olm-adding-pullsecret-catalogsource.adoc:
```diff
@@ -0,0 +1,46 @@
+    registryPoll:
+      interval: 30m
+----
+<1> Add a `spec.secrets` section and specify the secret.
```
It's better to add a `note` at the end: if the subscribed operator uses the private image, it will be failed to install. Because this `secret` will not be used for the Deployment. If the users want to this operator can be installed successfully. They have to add the auth info to the `pull-secret` of the `openshift-config` project. Related steps:
```console
$ oc extract secret/pull-secret -n openshift-config  --confirm
.dockerconfigjson
$ cat .dockerconfigjson | jq --compact-output '.auths["<registry>:<port>/<namespace>/"] |= . + {"auth":"xxx"}' > new_dockerconfigjson
$ oc set data secret/pull-secret -n openshift-config  --from-file=.dockerconfigjson=new_dockerconfigjson
secret/pull-secret data updated
```
Details: https://bugzilla.redhat.com/show_bug.cgi?id=1909992#c7

adellape commented:
@anik120 @jianzhangbjz @jeana-redhat I've updated this based on the latest feedback, and have moderately reworked the procedure, particularly the intro concept paragraphs. I've also changed the section title from "Adding pull secrets to catalog sources" to more broadly "Accessing images for Operators from private registries". Would you mind taking a fresh look again? Thank you!

http://file.rdu.redhat.com/~adellape/012021/olm_pullsecret/operators/admin/olm-managing-custom-catalogs.html#olm-accessing-images-private-registries_olm-managing-custom-catalogs

jeana-redhat commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,159 @@
+----
+<1> Add a `spec.secrets` section and specify any required secrets.
+
+. If any Operator or Operand images referenced by a subscribed Operator requires access to a private registry, add the authentication details to the global cluster pull secret in the `openshift-config` project.
```
I think subject-verb agreement is off here? "requires" is referring to "images", right? Might also help to use the complementizer her for clarity, e.g.
`If any Operator or Operand images that are referenced by a subscribed Operator require access [...]`

jeana-redhat commented on operators/admin/olm-managing-custom-catalogs.adoc:
```diff
@@ -19,6 +19,10 @@ include::modules/olm-understanding-operator-catalog-images.adoc[leveloffset=+1]
 include::modules/olm-creating-catalog-from-index.adoc[leveloffset=+2]
+.Additional resources
+
+* If your index image is hosted on a private registry and requires authentication, see xref:../../operators/admin/olm-managing-custom-catalogs.adoc#olm-accessing-images-private-registries_olm-managing-custom-catalogs[Adding pull secrets to catalog sources].
```
update link text to new title

jeana-redhat commented on operators/admin/olm-managing-custom-catalogs.adoc:
```diff
@@ -34,10 +38,17 @@ include::modules/olm-mirroring-package-manifest-catalog.adoc[leveloffset=+2]
 .Additional resources
 
 * xref:../../operators/operator_sdk/osdk-generating-csvs.adoc#olm-arch-os-support_osdk-generating-csvs[Architecture and operating system support for Operators]
+* If your catalog image is hosted on a private registry and requires authentication, see xref:../../operators/admin/olm-managing-custom-catalogs.adoc#olm-accessing-images-private-registries_olm-managing-custom-catalogs[Adding pull secrets to catalog sources].
```
update link text to new title

jeana-redhat commented on operators/admin/olm-managing-custom-catalogs.adoc:
```diff
@@ -34,10 +38,17 @@ include::modules/olm-mirroring-package-manifest-catalog.adoc[leveloffset=+2]
+.Additional resources
+
+* See xref:../../builds/creating-build-inputs.adoc#builds-secrets-overview_creating-build-inputs[What is a secret?] for more information on the types of secrets, including those used for registry credentials.
+* See xref:../../openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets[Using image pull secrets] for more details on the impact of updating the global cluster pull secret.
```
The link text here refers to the page but the link is to the section. Should the text be the section title instead?

adellape replied:
I went with the assembly title cuz I thought it was a little more useful in this context, vs saying something like "See 'Updating the global cluster pull secret' for more information on updating the global cluster pull secret." :shrug:

adellape replied:
Ended up going with:

```
* See xref:../../openshift_images/managing_images/using-image-pull-secrets.html#images-update-global-pull-secret_using-image-pull-secrets[Updating the global cluster pull secret] for more details on the impact of changing this secret.
```

jeana-redhat commented on operators/admin/olm-restricted-networks.adoc:
```diff
@@ -49,4 +49,6 @@ include::modules/olm-mirroring-catalog.adoc[leveloffset=+1]
 * xref:../../operators/operator_sdk/osdk-generating-csvs.adoc#olm-arch-os-support_osdk-generating-csvs[Architecture and operating system support for Operators]
 
 include::modules/olm-creating-catalog-from-index.adoc[leveloffset=+1]
+* If your index image is hosted on a private registry and requires authentication, see xref:../../operators/admin/olm-managing-custom-catalogs.adoc#olm-accessing-images-private-registries_olm-managing-custom-catalogs[Adding pull secrets to catalog sources].
```
update link text to new title

jeana-redhat reviewed (approved):
just some little comments, LGTM

anik120 reviewed (approved):
lgtm

jianzhangbjz commented:
/lgtm

adellape commented:
Reworked step 3 of the procedure to also include an option for linking pull secrets to the `default` SA of individual namespaces, instead of adding it to the global cluster pull secret. @jianzhangbjz Sorry to bug you again, but could you take a look at that updated last step?

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+
+The following types of images should be considered when determining whether Operators managed by OLM have appropriate pull access:
+
+Index or catalog images:: Index images and catalog images are containerized snapshots of an entire catalog. Index images reference bundle images. If an index or catalog image is hosted in a private registry, OperatorHub can use a secret to enable pull access.
```
```suggestion
CatalogSource images:: CatalogSource images are CatalogSources packaged as container images hosted in image registries. CatalogSources contain a collection of operator bundles. If CatalogSource is hosted in a private image registry, a secret can be used to enable pull access.
```

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+
+Index or catalog images:: Index images and catalog images are containerized snapshots of an entire catalog. Index images reference bundle images. If an index or catalog image is hosted in a private registry, OperatorHub can use a secret to enable pull access.
+
+Bundle images:: Operator bundle images are containerized Kubernetes manifests and Operator metadata that represent a unique version of an Operator. If any bundle images referenced in a catalog are hosted in one or more private registries, OperatorHub can use one or more secrets to enable pull access.
```
```suggestion
Bundle images:: Operator bundle images are bundle metadata/manifests packaged as container images that represent a unique version of an Operator. If any bundle images referenced in a CatalogSource are hosted in one or more private registries, one or more secrets can be used to enable pull access.
```

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+
+Bundle images:: Operator bundle images are containerized Kubernetes manifests and Operator metadata that represent a unique version of an Operator. If any bundle images referenced in a catalog are hosted in one or more private registries, OperatorHub can use one or more secrets to enable pull access.
+
+Operator and Operand images:: If an Operator subscribed from a catalog uses a private image, either for the Operator image itself or one of the Operand images it manages, the Operator will fail to install because the deployment will not have access to the required registry authentication. Referencing secrets in a catalog source does not enable OLM to place the secrets in target tenant namespaces.
```
```suggestion
Operator and Operand images:: If an Operator installed from a CatalogSource uses a private image, either for the Operator image itself or one of the Operand images it watches, the Operator will fail to install because the deployment will not have access to the required registry authentication. Referencing secrets in a CatalogSource does not enable OLM to place the secrets in target tenant namespaces that operands are installed in.
```

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+
+Operator and Operand images:: If an Operator subscribed from a catalog uses a private image, either for the Operator image itself or one of the Operand images it manages, the Operator will fail to install because the deployment will not have access to the required registry authentication. Referencing secrets in a catalog source does not enable OLM to place the secrets in target tenant namespaces.
++
+Instead, the authentication details can be added to the global cluster pull secret in the `openshift-config` namespace, which provides access to all namespaces on the cluster. Alternatively, if providing access to the entire cluster is not permissible, the pull secret can be added to the `default` service account of one or more namespaces to target individual tenants.
```
```suggestion
Instead, the authentication details can be added to the global cluster pull secret in the `openshift-config` namespace, which provides access to all namespaces on the cluster. Alternatively, if providing access to the entire cluster is not permissible, the pull secret can be added to the `default` service accounts of the targeted tenant namespaces.
```

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+.Prerequisites
+
+* At least one of the following hosted in a private registry:
+** An index image or catalog image.
```
```suggestion
** A CatalogSource.
```

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+
+* At least one of the following hosted in a private registry:
+** An index image or catalog image.
+** An Operator bundle image.
```
```suggestion
** An Operator bundle.
```

anik120 commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+* At least one of the following hosted in a private registry:
+** An index image or catalog image.
+** An Operator bundle image.
+** An Operator or Operand image.
```
```suggestion
** An Operator or Operand.
```

anik120 reviewed (commented):
@adellape the workflow looks good! Just had a few suggestions about the definitions.

jianzhangbjz commented on modules/olm-accessing-images-private-registries.adoc:
```diff
@@ -0,0 +1,173 @@
+    --from-file=.dockerconfigjson=new_dockerconfigjson
+----
+
+* To update individual namespaces, add your pull secret to the `default` service account in each namespace that you want to have access:
```
There are two problems here: 

1. The SA cannot link the `secret` from another namespace, so it requires the user to recreate the `secret` in this individual namespace. details: https://bugzilla.redhat.com/show_bug.cgi?id=1925377
2. this secret should be linked to the operator's SA, not the `default`. For example, etcd operator, it should be: `oc secret link etcd-operator secret-cs --for=pull`

adellape replied:
@jianzhangbjz Updated the step:

http://file.rdu.redhat.com/~adellape/012021/olm_pullsecret/operators/admin/olm-managing-custom-catalogs.html#olm-accessing-images-private-registries_olm-managing-custom-catalogs

jianzhangbjz replied:
@adellape I didn't find any updates on this pre-review doc.

jianzhangbjz commented:
The PR updates look good, but I still cannot find the updates in the above pre-review doc.
/lgtm

adellape commented:
/cherrypick enterprise-4.7
